src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Binders/BodyModelBinder.cs

```diff
@@ -132,7 +148,21 @@ public BodyModelBinder(IList<IInputFormatter> formatters, IHttpRequestStreamRead
                     return;
                 }
 
-                bindingContext.Result = ModelBindingResult.Success(model);
+                if (_isBindingRequired && model == null)
```

Comment:
I prefer the current location of the empty body check, mostly because the containing method (`ReadAsync()`) is `virtual` already. We don't have an example subclass that would return anything but `InputFormatterResult.NoData()` in that case. But, I agree that check should be changed to return `NoData` when the option says so.

If we want, say, `JsonInputFormatter` to also return `NoData` when the body contains only whitespace, we can do that now or later. Might be good to do it now because that's a case you're particularly concerned about @SteveSandersonMS.

Either way, pushing the empty body check into `JsonInputFormatter` etc. does not seem germane to the bug you're fixing here.